### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by bobvanderlinden**
# Motivation

Currently the Nix repository doesn't have a contributing document. This can help new contributors get started and explain what to expect.

As a response to [my post on discourse](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html) @fricklerhandwerk rightly suggested to create an initial version of the `CONTRIBUTING.md` document. This PR is the result.

I have based this on:

* https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md
* https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html

And I tried to refer to the manual whenever possible.

I took the liberty to 'fill in the blanks' by already linking to a code of conduct and filling in the steps that new contributors could take to help them get started.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:12]: We have [an initial CoC in the making](https://github.com/NixOS/rfcs/pull/114), but it's not merged yet. I wouldn't block this on the CoC, so let's remove it for now. We can always add it when it's there.

```suggestion
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:14]: Reading the [RoR guide on feature requests](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#what-about-feature-requests-questionmark), I think we should not actively encourage feature requests, but also not be as blunt and just close them away.

In my opinion, feature requests are somewhat a distraction, and while we need but haven't yet found a structured and predictably productive way to discuss them, there's also no need to further promote that ad hoc practice.

I suggest just not talking about feature suggestions at all, feels like the right middle ground until further notice.

```suggestion
## Report a bug
```

Would need fixups in list items following this

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:16]:
```suggestion
1. Check on the [GitHub issue tracker](https://github.com/NixOS/nix/issues) if your bug was already reported.
```

This is much shorter.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:22]:
```suggestion
   Make sure reported bugs can be reproduced easily.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:25]:
```suggestion
   The only way to ensure this, is to [work on the issue yourself](#making-changes-to-nix).
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:34]:
```suggestion
   There are many open pull requests that might already do what you intent to work on.
   You can use [labels](https://github.com/NixOS/nix/labels) to filter for relevant topics.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:38]:
```suggestion
3. Check the [Nix reference manual](https://nixos.org/manual/nix/unstable/contributing/hacking.html) for information on building Nix and running its tests.
```

We may want to link to unstable here, since this will be the most up-to-date version, and that's relevant for hacking.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:40]:
```suggestion
   For contributions to the command line interface, please check the [CLI guidelines](https://nixos.org/manual/nix/unstable/contributing/cli-guideline.html).
```

Again, possibly unstable?

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:44]:
```suggestion
5. [Create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) for your changes.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:50]:
```suggestion
   Nix maintainers follow a [structured process for reviews and design decisions](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol), which may or may not prioritise your work.
```

No need for value judgements on how busy we are.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:47]:
```suggestion
   * Link related issues in your pull request to inform interested parties and future contributors about your change.
     If your pull request closes one or multiple issues, note that in the description using `Closes: #<number>`, as it will then happen automatically when your change is merged.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:57]:
```suggestion
The Nix reference manual is hosted on https://nixos.org/manual/nix.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:58]:
```suggestion
The underlying source files are located in [`doc/manual/src`](./doc/manual/src).
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:59]: Meh, not sure about that. People will find out anyway, but I wouldn't exactly encourage them to. I agree though it's easier for many people to just do it via GitHub, and that lowers the barrier for contributions. I don't know.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:20]:
```suggestion
3. The issue templates will guide you in specifying your issue.
```

I think this is more specific and still brief. It's weird to say "something should be in an issue", at least it should be "issue description".

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:21]:
```suggestion
   The more complete the information you provide, the more likely it can be found by others and the more useful it is in the future.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:33]:
```suggestion
1. Check for [pull requests](https://github.com/NixOS/nix/pulls) that might already cover the contribution you are about to make.
```

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:53]:
```suggestion
7. If you need additional feedback or help to getting pull request into shape, ask other contributors using [@mentions](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#mentioning-people-and-teams).
```

Almost no one working on the code will look at Discourse, so it probably won't help. Also I wouldn't want that category to be flooded with "please review my PR" topics. Additionally, the discourse bot will make a comment every time someone links a PR on Discourse, and while often useful, this instruction would produce lots of unnecessary visual noise.

**Review by fricklerhandwerk — APPROVED** [review]
This looks really good, thanks a lot for taking time. Don't worry about the pile of comments, you should be able to just commit most of them if you agree they're reasonable.

File: CONTRIBUTING.md
Comment (bobvanderlinden) [reply CONTRIBUTING.md:53]: I'm not sure about this one. Having it be posted on discourse was what helped me get feedback. I do agree it can create visual noise of used too often.

Mentioning a contributor is an ok mechanism, but which one do you pick? A new contributor will likely just check who worked on a file last time and mention that person. There is a good chance that is a core contributor, which won't help to mention.

Is there a method/process to find an existing contributor that isn't on the core team?

For nixpkgs there is also https://discourse.nixos.org/t/prs-ready-for-review/3032/1380, but i wouldn't want to disturb that process with Nix PRs. And, it isn't a very nice process 😅

File: CONTRIBUTING.md
Comment (bobvanderlinden) [reply CONTRIBUTING.md:38]: I was kind of hesitant to link to unstable, because it might be that contributors use an older version of Nix, but indeed makes sense to still use unstable, as changes like changes to the build system and test setup will likely update `hacking` documentation as well.

File: CONTRIBUTING.md
Comment (bobvanderlinden) [reply CONTRIBUTING.md:59]: Indeed. I also mentioned this because of the markdown preview.

Maybe it is good to mention this is only for small changes, but refer to the hacking page in the manual for bigger changes?

Comment (fricklerhandwerk) [reply CONTRIBUTING.md:55]: Yes, sounds good.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [reply CONTRIBUTING.md]: These are three separate issues:
1. asking questions about the code and development
   
   This is what works best on Discourse, and this is what the Nix Category is for regarding this repo, and this is how you came here. It works, apparently, and I'd encourage it.

2. finding the right people to ask directly

   No obvious silver bullet, but git blame sounds like a good candidate. Yes, that's likely to turn out maintainers, but in the longer run this will diversify and already does.

3. asking for review and help on a given PR

   This should absolutely be done as @mentions, in place

File: CONTRIBUTING.md
Comment (bobvanderlinden) [reply CONTRIBUTING.md]: Done :+1:

I also squashes commits and rebased on master.

File: CONTRIBUTING.md
Comment (bobvanderlinden) [reply CONTRIBUTING.md]: Thanks, I've updated and now also formatted the documentation. I'm hoping this is enough information for contributors to go on. It should already help a bit.

Let me know if you have any other suggestions or what I can do to get this reviewed further and/or merged.

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting:
    - @edolstra: should be part of the hacking guide
    - @roberth: I like that it's in the standard location
    - @fricklerhandwerk: we can still wire it up later
        - @thufschmitt: relative links will break if we use symlink the file naively, needs more thought
     - decision: merge

**Comment by bobvanderlinden** [comment]
Nice, thanks!

**Pull request merged**
